torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -743,6 +755,145 @@ c10::optional<ScheduleHeuristic> tryMerge(
+  void mergeGroups(SegmentedGroup* a, SegmentedGroup* b, SegmentedGroup* ab) {
+    // TODO: avoid re-hashing
+    auto& ab_set = getSet(ab);
+    unionInto(ab, a);
+    unionInto(ab, b);
+    ab_set->erase(a);
+    ab_set->erase(b);
+    reach_map_.erase(a);
+    reach_map_.erase(b);
+    for (auto& it : reach_map_) {
+      if (it.second->count(a) || it.second->count(b)) {
+        it.second->insert(ab);
+        unionInto(it.first, ab);
+      }
+      it.second->erase(a);
+      it.second->erase(b);
+    }
+  }
+
+ private:
+  //! Collect initial reachable info using
+  //!  a work list algorithm through forward traversal
+  //!  a backward DFS would do the same
+  void computeReachability() {
+    GroupSet completed;
+    GroupSet work_list;
+
+    // Collect source nodes, we are guaranteed
+    //  a source node on a DAG and thus we can
+    //  be sure to make progress
+    std::copy_if(
+        segmented_fusion_->groups().begin(),
+        segmented_fusion_->groups().end(),
+        std::inserter(completed, completed.end()),
+        [](SegmentedGroup* g) { return g->producer_edges.empty(); });
+
+    // completed now only contain source nodes
+    //  they can go backward to nowhere
+    for (auto g : completed) {
+      addConsumersToWorkList(g, work_list);
+    }
+
+    while (!work_list.empty()) {
+      SegmentedGroup* to_update = nullptr;
+      for (auto g : work_list) {
+        if (std::all_of(
+                g->producer_edges.begin(),
+                g->producer_edges.end(),
+                [&completed](SegmentedEdge* e) {
+                  return completed.count(e->from);
+                })) {
+          // filter multi-edges
+          GroupSet producer_groups;
+          for (auto e : g->producer_edges) {
+            producer_groups.insert(e->from);
+          }
+
+          // populate all possible paths
+          // from producer backward, including
+          // the producer
+          for (auto p : producer_groups) {
+            insert(g, p);
+            unionInto(g, p);
+          }
+          to_update = g;
+          break;
+        }
+      }
+      if (to_update) {
+        addConsumersToWorkList(to_update, work_list);
+        work_list.erase(to_update);
+        completed.insert(to_update);
+      } else {
+        TORCH_INTERNAL_ASSERT(false, "unreachable");
+      }
+    }
+  }
+
+  void addConsumersToWorkList(SegmentedGroup* completed, GroupSet& work_list) {
+    for (auto e : completed->consumer_edges) {
+      // A consumer wouldn't have been worked before any of its producer
+      work_list.insert(e->to);
+    }
+  }
+
+  // Compute the union of two reachable maps and store in
+  //  in the entry of to
+  void unionInto(SegmentedGroup* to, SegmentedGroup* from) {
+    auto& from_set = *getSet(from);
+    for (auto g : from_set) {
+      insert(to, g);
+    }
+  }
+
+  //! Utility to access reachable map
+  GroupSetPtr& getSet(SegmentedGroup* key) {
+    auto& ptr = reach_map_[key];
```

Comment:
Is this safe? Is it guaranteed to have `key` in the map?

Reply:
I'd suggest using `at` unless the existence of a given key is extremely clear.

Reply:
Yes. Operator [] just creates an empty pointer if not matched, and the emptiness is checked in the following lines.

Reply:
You're right. I forgot unordered_map automatically creates a default value if key not found.